App/Containers/Contact.tsx

```diff
@@ -67,29 +80,59 @@ class ContactModal extends React.Component<Props> {
           >
             {this.props.displayName}
           </Text>
+          <Button
+            text={buttonText}
+            disabled={buttonDisabled}
+            onPress={this.props.isContact ? this.onRemove : this.onAdd}
+          />
         </View>
         <ScrollView style={styles.threadsList}>
           <Text style={styles.threadsTitle}>
             {this.props.threadThumbs.length > 0 ? 'Sharing in Groups:' : 'Not part of any shared groups'}
           </Text>
           {this.props.threadThumbs.map((thread, i) => (
-            <TouchableOpacity key={i} onPress={this.navigate(thread.id)}>
+            <TouchableOpacity key={i} onPress={this.navigateToThread(thread.id)}>
             <PhotoWithTextBox key={i} text={thread.name} photo={thread.thumb} />
             </TouchableOpacity>
           ))}
         </ScrollView>
       </View>
     )
   }
+
+  onRemove = () => {
+    this.props.removeContact(this.props.navigation.getParam('contact').address)
+  }
+
+  onAdd = () => {
+    this.props.addContact(this.props.navigation.getParam('contact'))
+  }
 }
 
 const mapStateToProps = (state: RootState, ownProps: NavigationScreenProps<NavProps>): StateProps => {
-  const username = ownProps.navigation.getParam('username')
-  const address = ownProps.navigation.getParam('address')
+  const contact = ownProps.navigation.getParam('contact')
+  const username = contact.name
+  const address = contact.address
+  // Check if this contact is already added
+  const isContact = state.contacts.contacts.some((c) => c.address === address)
+  // Check if this contact is currently being removed
+  const removing = Object.keys(state.contacts.removingContacts).indexOf(address) > -1
+  // Check if this contact is currently being added
+  const adding = Object.keys(state.contacts.addingContacts).indexOf(address) > -1
   return {
     displayName: username ? username : address.substring(0, 12),
-    threadThumbs: getThreadThumbs(state, address, 'name')
+    threadThumbs: getThreadThumbs(state, address, 'name'),
+    isContact,
+    removing,
+    adding
+  }
+}
+
+const mapDispatchToProps = (dispatch: Dispatch<RootAction>): DispatchProps => {
+  return {
+    removeContact: (address: string) => dispatch(contactsActions.removeContactRequest(address)),
+    addContact: (contact: IContact) => dispatch(contactsActions.addContactRequest(contact))
```

Comment:
One small, but nice improvement is using the optional `ownProps` parameter in `mapDispatchToProps`. Then you don't have to have `address: string` and `contact: IContact` as parameters to these prop functions... Inside the functions, you can just pull `address` and `contact` from `ownProps`.